### DISCUSSION TITLE
Messages stay in pending state forever

### DISCUSSION THREAD

**Discussion by bigon** [postalserver/postal#2669] [Help with using Postal]
Hello,

From time to time, some messages are staying in pending state forever.

In the `queued_messages` table, the messages are lock for several hours and the `retry_after` column is set to NULL

I'm not 100% sure what's happening here, I see that the requeuer container has exited (code 1) and I also see rabbitmq connection issues. (I'm not whether the connection issue I see in the logs are causing the container to exit 1 or if some issues in the code causes the errors about the connections and then the container crashes)

So two things:

1. Shouldn't be some kind of rollback of the lock if there is an issue with the rabbitmq connection and/or if a container is killed (of course that shouldn't happen, but still)
2. Shouldn't be some kind of house keeping removing the locks if a messages is locked for more than X hours/days/...

Version of postal 2.1.1

Rabbitmq warning/errors I see in the requeuer logs:

```
rake aborted!
Bunny::ConnectionClosedError: Trying to send frame through a closed connection. Frame is [#<AMQ::Protocol::MethodFrame:0x0000556a09277928 @payload="\x00<\x00(\x00\x00\x00\x11deliver-jobs-main\x00", @channel=1>, #<AMQ::Protocol::HeaderFrame:0x0000556a09277180 @payload="\x00<\x00\x00\x00\x00\x00\x00\x00\x00\x00\\\x98\x00\x18application/octet-stream\x01\x00", @channel=1>, #<AMQ::Protocol::BodyFrame:0x0000556a09277130 @payload="{\"params\":{\"id\":69915940},\"class_name\":\"UnqueueMessageJob\",\"id\":\"CHC3ID5XGL\",\"queue\":\"main\"}", @channel=1>]
/usr/local/bundle/gems/bunny-2.14.4/lib/bunny/session.rb:1110:in `block in send_frameset'
/usr/local/bundle/gems/bunny-2.14.4/lib/bunny/channel.rb:1460:in `synchronize'
/usr/local/bundle/gems/bunny-2.14.4/lib/bunny/session.rb:1103:in `send_frameset'
/usr/local/bundle/gems/bunny-2.14.4/lib/bunny/channel.rb:568:in `basic_publish'
/usr/local/bundle/gems/bunny-2.14.4/lib/bunny/exchange.rb:141:in `publish'
/usr/local/bundle/gems/bunny-2.14.4/lib/bunny/queue.rb:280:in `publish'
/opt/postal/app/lib/postal/job.rb:38:in `queue'
/opt/postal/app/app/models/queued_message.rb:46:in `queue'
/usr/local/bundle/gems/activerecord-5.2.6.2/lib/active_record/relation/delegation.rb:71:in `each'
/usr/local/bundle/gems/activerecord-5.2.6.2/lib/active_record/relation/delegation.rb:71:in `each'
/opt/postal/app/app/models/queued_message.rb:119:in `requeue_all'
/opt/postal/app/lib/postal/message_requeuer.rb:11:in `block in run'
/opt/postal/app/lib/postal/message_requeuer.rb:9:in `loop'
/opt/postal/app/lib/postal/message_requeuer.rb:9:in `run'
/opt/postal/app/lib/tasks/postal.rake:18:in `block (2 levels) in <top (required)>'
/opt/postal/app/bin/bundle:3:in `load'
/opt/postal/app/bin/bundle:3:in `<main>'
Tasks: TOP => postal:requeuer
(See full trace by running task with --trace)
```

**Comment by bigon** [comment]
Note that, I see both messages that were generated at the time of the crash/issue with the rabbitmq connection AND later messages.

So it seems that there might be 2 issues here

**Comment by willpower232** [comment]
If you have a look at `postal status` does it show that the workers have restarted? Not sure if you'll see that but would be a good indication.

The queued_messages table is probably locked by a worker having a bad time halfway through so the locked columns may need clearing manually.

If you're running everything on one server, is it possible that you're running out of memory to keep everything going?

**Reply by bigon** [reply]
> The queued_messages table is probably locked by a worker having a bad time halfway through so the locked columns may need clearing manually.

The fact that that the messages are locked forever and need a manual modification of the table is IMVHO a problem, messages should be retried later (messages staying locked for more than X hours should show definitely be an issue)

And I don't think that the crash of the requeuer is a memory issue, as you can see there is a trace caused by the rabbitMQ client library not being happy that the connection has been closed (Bunny::ConnectionClosedError: Trying to send frame through a closed connection)

**Reply by willpower232** [reply]
> The fact that that the messages are locked forever and need a manual modification of the table is IMVHO a problem, messages should be retried later (messages staying locked for more than X hours should show definitely be an issue)

It would be good if there were some automatic resolution for this but as far as I know from the issues and discussions here, it hasn't come up very often.

If you have a look in the database, you can see columns like "locked_by" or something which is supposed to reference the process of the worker responsible for the message, if you know the message should have been sent then you can clear the columns by yourself and see if the message gets picked up for sending by the current set of workers.

**Comment by Jank1310** [comment]
Had the same issue. Solved it by clearing the locked_by and locked_at columns of the messages. The reason is a shutdown of the server (postal stop) while processing messages.

**Reply by adamcooke** [reply]
This should be handled better in Postal v3.
